From 077ab17a1eed1ceaed43e193f03aa7daefe95485 Mon Sep 17 00:00:00 2001
From: Alexander Duyck <alexander.h.duyck@redhat.com>
Date: Fri, 31 Oct 2014 19:06:33 -0400
Subject: [net] gre: Use inner mac length when computing tunnel length

Message-id: <20141031190633.8251.13132.stgit@ahduyck-vm-rhel7>
Patchwork-id: 98656
O-Subject: [RHEL7.1 net PATCH 4/7] gre: Use inner mac length when computing tunnel length
Bugzilla: 1151886 1152368
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>

Bugzilla: 1151886
Brew Build: 8179440

Backport of upstream commit: 14051f0452a2c26a3f4791e6ad6a435e8f1945ff

 commit 14051f0452a2c26a3f4791e6ad6a435e8f1945ff
 Author: Tom Herbert <therbert@google.com>
 Date:   Thu Oct 30 08:40:56 2014 -0700

    gre: Use inner mac length when computing tunnel length

    Currently, skb_inner_network_header is used but this does not account
    for Ethernet header for ETH_P_TEB. Use skb_inner_mac_header which
    handles TEB and also should work with IP encapsulation in which case
    inner mac and inner network headers are the same.

    Tested: Ran TCP_STREAM over GRE, worked as expected.

    Signed-off-by: Tom Herbert <therbert@google.com>
    Acked-by: Alexander Duyck <alexander.h.duyck@redhat.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Alexander Duyck <alexander.h.duyck@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/gre_offload.c b/net/ipv4/gre_offload.c
index 1e35e35..5771ffc 100644
--- a/net/ipv4/gre_offload.c
+++ b/net/ipv4/gre_offload.c
@@ -51,7 +51,7 @@ static struct sk_buff *gre_gso_segment(struct sk_buff *skb,
 
  greh = (struct gre_base_hdr *)skb_transport_header(skb);
 
- ghl = skb_inner_network_header(skb) - skb_transport_header(skb);
+ ghl = skb_inner_mac_header(skb) - skb_transport_header(skb);
  if (unlikely(ghl < sizeof(*greh)))
   goto out;
 
-- 
1.7.1